complex/ft_msg.c

```diff
@@ -542,30 +556,80 @@ int ft_post_recv_bufs(void)
 	return 0;
 }
 
-int ft_recv_msg(void)
+int ft_recv_n_msg(int n)
 {
-	int credits, ret;
+	int credits, ret, recved = 0;
 
+post:
 	if (ft_rx_ctrl.credits > (ft_rx_ctrl.max_credits >> 1)) {
 		ret = ft_post_recv_bufs();
 		if (ret)
 			return ret;
 	}
 
-	credits = ft_rx_ctrl.credits;
-	do {
+	while (recved < n) {
+		credits = ft_rx_ctrl.credits;
 		ret = ft_comp_rx(FT_COMP_TO);
 		if (ret)
 			return ret;
+
+		recved += (ft_rx_ctrl.credits - credits);
```

Comment:
I think something is still off here.  Suppose the first call to ft_recv_n_msg() asks for 10 messages, but the code finds 11.  The next time, the caller asks for 10 more messages.  There's no way to track that one of those 10 was found by the prior call.  If this call is going to ask for exactly n completions, it should only process that many.